services/src/main/java/org/keycloak/protocol/oid4vc/issuance/signing/JwtProofBasedSigningService.java

```diff
@@ -0,0 +1,208 @@
+                        throw new VCIssuerException("Wrong proof type");
+                    }
+
+                    return Optional.of(proof);
+                });
+    }
+
+    private JWSInput getJwsInput(Proof proof) throws JWSInputException {
+        return new JWSInput(proof.getJwt());
+    }
+
+    /**
+     * As we limit accepted algorithm to the ones listed by the issuer, we can omit checking for "none"
+     * The Algorithm enum class does not list the none value anyway.
+     *
+     * @param vcIssuanceContext
+     * @param jwsHeader
+     * @throws VCIssuerException
+     */
+    private void validateJwsHeader(VCIssuanceContext vcIssuanceContext, JWSHeader jwsHeader) throws VCIssuerException {
+        Optional.ofNullable(jwsHeader.getAlgorithm())
+                .orElseThrow(() -> new VCIssuerException("Missing jwsHeader claim alg"));
+
+        // As we limit accepted algorithm to the ones listed by the server, we can omit checking for "none"
+        // The Algorithm enum class does not list the none value anyway.
+        Optional.ofNullable(vcIssuanceContext.getCredentialConfig())
+                .map(SupportedCredentialConfiguration::getProofTypesSupported)
+                .map(ProofTypesSupported::getJwt)
+                .map(ProofTypeJWT::getProofSigningAlgValuesSupported)
+                .filter(supportedAlgs -> supportedAlgs.contains(jwsHeader.getAlgorithm().name()))
+                .orElseThrow(() -> new VCIssuerException("Proof signature algorithm not supported: " + jwsHeader.getAlgorithm().name()));
+
+        Optional.ofNullable(jwsHeader.getType())
+                .filter(type -> Objects.equals(PROOF_JWT_TYP, type))
+                .orElseThrow(() -> new VCIssuerException("JWT type must be: " + PROOF_JWT_TYP));
+
+        // KeyId shall not be present alongside the jwk.
+        Optional.ofNullable(jwsHeader.getKeyId())
+                .ifPresent(keyId -> {
+                    throw new VCIssuerException("KeyId not expected in this JWT. Use the jwk claim instead.");
+                });
+    }
+
+    private void validateProofPayload(VCIssuanceContext vcIssuanceContext, AccessToken proofPayload) throws VCIssuerException {
+        // azp is the id of the client, as mentioned in the access token used to request the credential.
+        // Token provided from user is obtained with a clientId that support the oidc login protocol.
+        // oid4vci client doesn't. But it is the client needed at the credential endpoint.
+        //        String azp = vcIssuanceContext.getAuthResult().getToken().getIssuedFor();
+        //        Optional.ofNullable(proofPayload.getIssuer())
+        //                .filter(proofIssuer -> Objects.equals(azp, proofIssuer))
+        //                .orElseThrow(() -> new VCIssuerException("Issuer claim must be null for preauthorized code else the clientId of the client making the request: " + azp));
+
```

Comment:
> Could you confirm the following?
> 
> A client (1) that sends a token request and receives an access token is the one whose protocol is oidc.

Yes. To obtain a token, i need an oidc client. Or do you think i can use an oid4vci client to obtain a token?

> A client (2) that sends a credential request and receives a VC is the one whose protocol is oid4vci.

Yes. Unless we have a way of issuing a token with an oidc client.

> Both shold be the same, but Keycloak does not do that ( azp in an access token shows the client (1) while a sender of Key binding proof is the client (2)).

This is why i did not check the azp here and left it commented. We need a concept.

> To get around it, the client (1)'s access token includes audiences that includes the client (2), and the method check it.

Ok. I will try this. As client access token can have many audiences. Thank you.

> If those above are right, IMO, it might be preferable to require only one client for OID4VCI.
> 
> For example, [POTENTIAL Interop Event / Track 2 / Full Profile](https://www.authlete.com/developers/oid4vci/#45-potential-interop-event--track-2--full-profile) requires DPoP Proof JWT on Token Endpiont and Credential Endpoint, which means that a sender of a token request should be the sender of a credential request. I am afraid that the current Keycloak's treatment of clients for OID4VCI does not realize such the use-case.

Correct. I will try the approach with many audiences.

> 
> Also, for tracking of VC issuance for audit, it might be preferable that one client go through all phases from getting a credential offer, an access token, and a VC.

Yes.

I will create a separate ticket to implement this: https://github.com/keycloak/keycloak/issues/31496